prometheus-proc/src/Prometheus/Metric/Proc.hs

```diff
@@ -0,0 +1,234 @@
+        "Total user and system CPU time spent in seconds."
+        CounterType
+        ( fromTicks ( utime + stime ) )
+
+    process_start_time_seconds = do
+      btime <-
+        mbtime
+
+      return
+        ( metric
+            "process_start_time_seconds"
+            "Start time of the process since unix epoch in seconds."
+            GaugeType
+            ( fromIntegral btime + fromTicks starttime )
+        )
+
+    process_virtual_memory_bytes =
+      metric
+        "process_virtual_memory_bytes"
+        "Virtual memory size in bytes."
+        GaugeType
+        vsize
+
+    process_resident_memory_bytes =
+      metric
+        "process_resident_memory_bytes"
+        "Resident memory size in bytes."
+        GaugeType
+        ( rss * sysconfPageSize )
+
+    metric metricName metricHelp metricType value =
+      SampleGroup
+        Info{..}
+        metricType
+        [ Sample
+            metricName
+            []
+            ( fromString ( show value ) )
+        ]
+
+
+-- | Convert a number of clock ticks into the corresponding duration in seconds.
+fromTicks :: Int -> Double
+fromTicks ticks =
+  fromIntegral ticks / fromIntegral clk_tck
+
+
+{-|
+
+Returns the current boot time in seconds since Unix epoch. This is a Maybe
+as we might not to be able to successfully parse this information out of
+@/proc/stat@.
+
+'unsafePerformIO' is used as this value does not change during the
+execution of the program, so this gives us a lightweight cache for this
+value.
+
+-}
+{-# NOINLINE mbtime #-}
+mbtime :: Maybe Int
+mbtime = unsafePerformIO $ do
+  fmap ( \( _, a, _ ) -> a ) . RE.findFirstInfix ( "btime " *> RE.decimal )
+    <$> readFile "/proc/stat"
+
+
+-- | Specific metrics from @/proc/xyz/stat@ that we are interested in.
+data ProcStat = ProcStat
+  { utime :: Int
+    -- ^ Amount of time that this process has been scheduled in user mode,
+    -- measured in clock ticks (divide by sysconf(_SC_CLK_TCK)).
+  , stime :: Int
+    -- ^ Amount of time that this process has been scheduled in kernel mode,
+    -- measured in clock ticks (divide by sysconf(_SC_CLK_TCK)).
+  , starttime :: Int
+    -- ^ The time the process started after system boot. In kernels before Linux
+    -- 2.6, this value was expressed in jiffies. Since Linux 2.6, the value is
+    -- expressed in clock ticks (divide by sysconf(_SC_CLK_TCK)).
+  , vsize :: Int
+    -- ^ Virtual memory size in bytes.
+  , rss :: Int
+    -- ^ Resident Set Size: number of pages the process has in real memory. This
+    -- is just the pages which count toward text, data, or stack space. This
+    -- does not include pages which have not been demand-loaded in, or which are
+    -- swapped out.
+  }
+  deriving
+    ( Show )
+
+
+{-|
+
+A regular expression for parsing @/proc/xyz/stat@. See
+@man 5 proc@ for more information on the format of this file:
+https://linux.die.net/man/5/proc.
+
+-}
+parseProcStat :: RE.RE Char ProcStat
+parseProcStat =
```

Comment:
If there were any weird cases you encountered while writing this it might be nice to include them as unit tests. If this needs to be tweaked in the future it would help ensure it doesn't regress on things you've already encountered.